spacelift/resource_stack_dependency_reference.go

```diff
@@ -0,0 +1,179 @@
+package spacelift
+
+import (
+	"context"
+	"path"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/spacelift-io/terraform-provider-spacelift/spacelift/internal"
+	"github.com/spacelift-io/terraform-provider-spacelift/spacelift/internal/structs"
+)
+
+func resourceStackDependencyReference() *schema.Resource {
+	return &schema.Resource{
+		Description: "" +
+			"`spacelift_stack_dependency_reference` represents a Spacelift **stack dependency reference** - " +
+			"a reference matches a stack's output to another stack's input. It is similar to an environment variable " +
+			"(`spacelift_environment_variable`), except that value is provided by another stack's output.",
+
+		CreateContext: resourceStackDependencyReferenceCreate,
+		ReadContext:   resourceStackDependencyReferenceRead,
+		UpdateContext: resourceStackDependencyReferenceUpdate,
+		DeleteContext: resourceStackDependencyReferenceDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"stack_dependency_id": {
+				Type:        schema.TypeString,
+				Description: "Immutable ID of stack dependency",
+				Required:    true,
+				ForceNew:    true,
+			},
+			"output_name": {
+				Type:        schema.TypeString,
+				Description: "Name of the output of stack to depend on",
+				Required:    true,
+			},
+			"input_name": {
+				Type:        schema.TypeString,
+				Description: "Name of the input of the stack dependency reference",
+				Required:    true,
+			},
+		},
+	}
+}
+
+func resourceStackDependencyReferenceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var query struct {
+		StackDependencyReference structs.StackDependencyReference `graphql:"stackDependenciesAddReference(stackDependencyID: $stackDependencyID, reference: $reference)"`
+	}
+
+	idParts, diags := getStackDependencyIDParts(d)
+	if diags != nil {
+		return diags
+	}
+
+	variables := map[string]interface{}{
+		"stackDependencyID": toID(idParts[1]),
+		"reference": structs.StackDependencyReferenceInput{
+			OutputName: toString(d.Get("output_name")),
+			InputName:  toString(d.Get("input_name")),
+			Type:       toString("ENVIRONMENT_VARIABLE"),
+		},
+	}
+
+	if err := meta.(*internal.Client).Mutate(ctx, "StackDependenciesAddReference", &query, variables); err != nil {
+		return diag.Errorf("could not create stack dependency reference: %s", err)
+	}
+
+	d.SetId(path.Join(idParts[0], idParts[1], query.StackDependencyReference.ID))
```

Comment:
hm, I think we have reference ULIDs, right? That'd be a good ID.

Reply:
Yes, `query.StackDependencyReference.ID` is ULID. But we also need `stack_dependency_id` and `stack_id`.
We querying: `stack($stack_id) { dependency($stack_dependency_id) { reference($id) { ... } } } }`. 
Which is similar to querying dependency (but one layer less). 

We could add a direct query (like `stack_dependency_reference($id)` without nested queries), but I saw here many places where we use this id format (`idPart[0]/idPart[1]/...`). 

Btw, we also verify a stack or dependency exists; a stack contains dependency, etc... Skipping stack_id and dependency_id, we can't do these checks.

Reply:
ah nevermind, right